src/Corax/Querying/Matches/TermMatch.cs

```diff
@@ -277,109 +278,96 @@ static QueryInspectionNode InspectFunc(ref TermMatch term)
             {
-                _bm25Relevance = isBoosting 
-                    ? Bm25Relevance.Small(indexSearcher, itemsCount, ctx, itemsCount, termRatioToWholeCollection) 
+                _bm25Relevance = isBoosting
+                    ? Bm25Relevance.Small(indexSearcher, itemsCount, ctx, itemsCount, termRatioToWholeCollection)
                     : default,
                 _current = 0,
                 _containerItem = containerItem,
                 _containerReader = reader
             };
         }
-        
-        public static TermMatch YieldSet(IndexSearcher indexSearcher, ByteStringContext ctx, PostingList postingList, double termRatioToWholeCollection, bool isBoosting, bool useAccelerated = true)
+
+        public static TermMatch YieldSet(IndexSearcher indexSearcher, ByteStringContext ctx, PostingList postingList, double termRatioToWholeCollection, bool isBoosting,
+            bool useAccelerated = true)
         {
             [SkipLocalsInit]
             [MethodImpl(MethodImplOptions.AggressiveInlining)]
-            static int AndWithFunc<TBoostingMode>(ref TermMatch term, Span<long> buffer, int matchesCount) where TBoostingMode : IBoostingMarker
+            static int AndWithFunc<TBoostingMode>(ref TermMatch term, Span<long> matches, int matchesCount) where TBoostingMode : IBoostingMarker
             {
                 if (matchesCount == 0)
                     return 0;
                 
-                int resultCursor = 0;
-                int bufferCursor = 0;
-                
                 var it = term._set;
-                ref long matchesBuffer = ref MemoryMarshal.GetReference(buffer);
-
-                var matchesLowest = Unsafe.Add(ref matchesBuffer, 0);
-                var matchesHighest = Unsafe.Add(ref matchesBuffer, matchesCount - 1);
-                
-                if (it.Seek(matchesBuffer - 1) == false)
+                // There is no element equal to or greater than `min(matches)-1`. No matches possible anyway.
+                if (it.Seek(EntryIdEncodings.PrepareIdForSeekInPostingList(matches[0] - 1)) == false)
                     return 0;
 
-                Span<long> postingListItems = stackalloc long[1024];
-                ref var postingListBuffer = ref MemoryMarshal.GetReference(postingListItems);
-                
-                //Maximum value we can match.
-                var maxValidValue = Unsafe.Add(ref matchesBuffer, matchesCount - 1) + 1;
-                var currentMatchesDocument = Unsafe.Add(ref matchesBuffer, bufferCursor);
-
-                while (resultCursor < matchesCount && it.Fill(postingListItems, out var read, maxValidValue) && read > 0)
+                var maxValidValue = EntryIdEncodings.PrepareIdForPruneInPostingList(matches[^1] + 1);
+                var postingListBuffer = stackalloc long[1024];
+                fixed (long* matchesPtr = matches)
```

Comment:
Why are we using `fixed` here? Unsafe pointer arithmetic create far better code and we made a conscious effort to remove it as much as we could. Unless this prototype code to test stuff.

Reply:
@maciejaszyk ?